File: frontend/src/api/reports.ts
import apiClient from './client'

export interface ProjectSummary {
  total_tickets: number
  open_tickets: number
  in_progress_tickets: number
  done_tickets: number
  overdue_tickets: number
  total_story_points: number
  completed_story_points: number
  by_priority: Record<string, number>
  by_type: Record<string, number>
}

export interface VelocityEntry {
  sprint_id: string
  sprint_name: string
  velocity: number
}

export interface VelocityReport {
  entries: VelocityEntry[]
  average: number
}

export interface BurndownPoint {
  date: string
  remaining: number
  ideal: number
}

export interface BurndownReport {
  sprint_id: string
  points: BurndownPoint[]
}

export interface CfdDay {
  date: string
  todo: number
  in_progress: number
  done: number
}

export interface CumulativeFlowReport {
  project_id: string
  days: CfdDay[]
}

export interface CycleTimeEntry {
  ticket_id: string
  ticket_key: string | null
  title: string
  cycle_time_hours: number
}

export interface CycleTimeReport {
  entries: CycleTimeEntry[]
  average_hours: number
  median_hours: number
}

export interface SprintReportTicket {
  ticket_id: string
  ticket_key: string | null
  title: string
  story_points: number
  completed: boolean
  priority: string
  ticket_type: string
}

export interface SprintReportSummary {
  total_tickets: number
  completed_tickets: number
  incomplete_tickets: number
  total_story_points: number
  completed_story_points: number
  incomplete_story_points: number
  completion_rate: number
}

export interface SprintReport {
  sprint_id: string
  sprint_name: string
  status: string
  start_date: string | null
  end_date: string | null
  summary: SprintReportSummary
  tickets: SprintReportTicket[]
}

export async function getSprintReport(projectId: string, sprintId: string): Promise<SprintReport> {
  const { data } = await apiClient.get<SprintReport>(`/projects/${projectId}/sprints/${sprintId}/report`)
  return data
}

export async function getProjectSummary(projectId: string): Promise<ProjectSummary> {
  const { data } = await apiClient.get<ProjectSummary>(`/projects/${projectId}/reports/summary`)
  return data
}

export async function getVelocityReport(projectId: string): Promise<VelocityReport> {
  const { data } = await apiClient.get<VelocityReport>(`/projects/${projectId}/reports/velocity`)
  return data
}

export async function getBurndownReport(sprintId: string): Promise<BurndownReport> {
  const { data } = await apiClient.get<BurndownReport>(`/sprints/${sprintId}/reports/burndown`)
  return data
}

export async function getCumulativeFlowReport(
  projectId: string,
  startDate: string,
  endDate: string,
): Promise<CumulativeFlowReport> {
  const { data } = await apiClient.get<CumulativeFlowReport>(
    `/projects/${projectId}/reports/cumulative-flow`,
    { params: { start_date: startDate, end_date: endDate } },
  )
  return data
}

export async function getCycleTimeReport(
  projectId: string,
  params?: { start_date?: string; end_date?: string },
): Promise<CycleTimeReport> {
  const { data } = await apiClient.get<CycleTimeReport>(
    `/projects/${projectId}/reports/cycle-time`,
    { params },
  )
  return data
}

File: backend/app/api/v1/endpoints/reports.py
from __future__ import annotations

import csv
import io
from datetime import date
from uuid import UUID

from fastapi import APIRouter, Depends, Query
from fastapi.responses import StreamingResponse
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_user, get_db
from app.core.permissions import ProjectRole, require_project_role
from app.models.user import User
from app.schemas.report import (
    BurndownReport,
    CumulativeFlowReport,
    CycleTimeReport,
    ProjectSummary,
    SprintReport,
    VelocityReport,
)
from app.services import report_service

router = APIRouter(tags=["reports"])


def _csv_response(rows: list[dict], filename: str) -> StreamingResponse:
    if not rows:
        buf = io.StringIO("")
    else:
        buf = io.StringIO()
        writer = csv.DictWriter(buf, fieldnames=list(rows[0].keys()))
        writer.writeheader()
        writer.writerows(rows)
    buf.seek(0)
    return StreamingResponse(
        buf,
        media_type="text/csv",
        headers={"Content-Disposition": f'attachment; filename="{filename}"'},
    )


@router.get(
    "/projects/{project_id}/reports/summary",
    response_model=ProjectSummary,
)
async def get_project_summary(
    project_id: UUID,
    user: User = Depends(get_current_user),
    _role: ProjectRole = require_project_role(ProjectRole.GUEST),
    db: AsyncSession = Depends(get_db),
):
    return await report_service.get_project_summary(db, project_id)


@router.get(
    "/projects/{project_id}/reports/velocity",
    response_model=VelocityReport,
)
async def get_velocity(
    project_id: UUID,
    user: User = Depends(get_current_user),
    _role: ProjectRole = require_project_role(ProjectRole.GUEST),
    db: AsyncSession = Depends(get_db),
):
    return await report_service.get_velocity_report(db, project_id)


@router.get(
    "/projects/{project_id}/sprints/{sprint_id}/report",
    response_model=SprintReport,
)
async def get_sprint_report(
    project_id: UUID,
    sprint_id: UUID,
    user: User = Depends(get_current_user),
    _role: ProjectRole = require_project_role(ProjectRole.GUEST),
    db: AsyncSession = Depends(get_db),
):
    return await report_service.get_sprint_report(db, sprint_id)


@router.get(
    "/sprints/{sprint_id}/reports/burndown",
    response_model=BurndownReport,
)
async def get_burndown(
    sprint_id: UUID,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await report_service.get_burndown_report(db, sprint_id)


@router.get(
    "/projects/{project_id}/reports/cumulative-flow",
    response_model=CumulativeFlowReport,
)
async def get_cumulative_flow(
    project_id: UUID,
    start_date: date = Query(...),
    end_date: date = Query(...),
    user: User = Depends(get_current_user),
    _role: ProjectRole = require_project_role(ProjectRole.GUEST),
    db: AsyncSession = Depends(get_db),
):
    return await report_service.get_cumulative_flow(
        db, project_id, start_date, end_date,
    )


@router.get(
    "/projects/{project_id}/reports/cycle-time",
    response_model=CycleTimeReport,
)
async def get_cycle_time(
    project_id: UUID,
    start_date: date | None = Query(None),
    end_date: date | None = Query(None),
    user: User = Depends(get_current_user),
    _role: ProjectRole = require_project_role(ProjectRole.GUEST),
    db: AsyncSession = Depends(get_db),
):
    return await report_service.get_cycle_time_report(
        db, project_id, start_date=start_date, end_date=end_date,
    )


# ── CSV exports ──────────────────────────────────────────────────


@router.get("/projects/{project_id}/reports/velocity/csv")
async def export_velocity_csv(
    project_id: UUID,
    user: User = Depends(get_current_user),
    _role: ProjectRole = require_project_role(ProjectRole.GUEST),
    db: AsyncSession = Depends(get_db),
):
    data = await report_service.get_velocity_report(db, project_id)
    rows = [
        {"sprint_name": e["sprint_name"], "velocity": e["velocity"]}
        for e in data["entries"]
    ]
    return _csv_response(rows, "velocity.csv")


@router.get("/projects/{project_id}/reports/cycle-time/csv")
async def export_cycle_time_csv(
    project_id: UUID,
    start_date: date | None = Query(None),
    end_date: date | None = Query(None),
    user: User = Depends(get_current_user),
    _role: ProjectRole = require_project_role(ProjectRole.GUEST),
    db: AsyncSession = Depends(get_db),
):
    data = await report_service.get_cycle_time_report(
        db, project_id, start_date=start_date, end_date=end_date,
    )
    rows = [
        {"ticket_key": e["ticket_key"], "title": e["title"], "cycle_time_hours": e["cycle_time_hours"]}
        for e in data["entries"]
    ]
    return _csv_response(rows, "cycle-time.csv")


@router.get("/projects/{project_id}/reports/cumulative-flow/csv")
async def export_cfd_csv(
    project_id: UUID,
    start_date: date = Query(...),
    end_date: date = Query(...),
    user: User = Depends(get_current_user),
    _role: ProjectRole = require_project_role(ProjectRole.GUEST),
    db: AsyncSession = Depends(get_db),
):
    data = await report_service.get_cumulative_flow(
        db, project_id, start_date, end_date,
    )
    rows = [
        {"date": d["date"], "todo": d["todo"], "in_progress": d["in_progress"], "done": d["done"]}
        for d in data["days"]
    ]
    return _csv_response(rows, "cumulative-flow.csv")
